observations/download_utils.py

```diff
@@ -0,0 +1,541 @@
+    sys.stdout.write(o_str % (
+                     humanize_size(progress_size),
+                     'Unknown',
+                     humanize_size(speed),
+                     humanize_time(duration)))
+    sys.stdout.flush()
+
+
+def normal_download(url, file_path, hash_true=None, timeout=10):
+  """start download of a file. No pause/resume capability
+    Args:
+      url: str
+        URL string to the file
+      file_path: str
+        full path to filename to be saved locally
+      hash_true: str [None]
+        true hash_string of the requested file
+      timeout: int [10]
+        timeout in secs
+    Returns:
+      bool. True if file is successfully downloaded
+  """
+  file_size = get_file_size(url, timeout)
+  start_time = 0
+  if check_file_downloaded(file_path, file_size):
+    print('File [{}] with same size already downloaded & '
+          'exists in the provided path'.format(file_path))
+    return True
+  else:
+    # define callback to print progress
+    def _reporthook(count, block_size, total_size):
+      global start_time
+      if count == 0:
+        start_time = time.time()
+        return
+      progress_size = min(int(count * block_size), file_size)
+      if (total_size - progress_size < block_size) or count % 100 == 0:
+        duration = time.time() - start_time
+        remaining_size = total_size - progress_size
+        speed = int(progress_size / duration)
+        eta = int(remaining_size / speed)
+        eta = eta if eta >= 0 else 0
+        percent = min(int(count * block_size * 100 / total_size), 100)
+        o_str = '\r>> [%s/%s] %d%% @%s/s,[%s remaining,' \
+                ' %s elapsed]      '
+        sys.stdout.write(o_str % (
+                         humanize_size(progress_size),
+                         humanize_size(file_size),
+                         percent,
+                         humanize_size(speed),
+                         humanize_time(eta),
+                         humanize_time(duration)))
+        sys.stdout.flush()
+
+    try:
+      print('>> Downloading {}' . format(file_path))
+      urllib.request.urlretrieve(url, file_path, _reporthook)
+      if file_size == os.path.getsize(file_path):
+        if hash_true is not None and not hashes_match(file_path,
+                                                      hash_true):
+          raise Exception('Error validating the file '
+                          'against its MD5 hash')
+        if hash_true is None:
+          print("\nURL {} downloaded to {} ".format(url, file_path))
+        else:
+          print("\nURL {} downloaded to {} and hash verified "
+                "succesfully".format(url, file_path))
+        return True
+    except:
+      if os.path.exists(file_path):
+        os.remove(file_path)
+      raise
+
+
+def download_file(url, file_path, hash_true=None, resume=True,
+                  block_size=1024 * 1024):
+  """Start or resume partially downloaded file from specified URL
```

Comment:
All docstring comments should end with a period including in the `Args:` and `Returns:` sections.